src/cpu/aarch64/matmul/acl_matmul.hpp

```diff
@@ -0,0 +1,155 @@
+
+        using cpu_matmul_pd_t::cpu_matmul_pd_t;
+
+        DECLARE_COMMON_PD_T("matmul:acl", acl_matmul_t, USE_GLOBAL_SCRATCHPAD);
+
+        status_t init(engine_t *engine) {
+            using smask_t = primitive_attr_t::skip_mask_t;
+            bool ok = src_md()->data_type == data_type::f32
+                    && weights_md()->data_type == data_type::f32
+                    && desc()->accum_data_type == data_type::f32
+                    && dst_md()->data_type == data_type::f32
+                    && platform::has_data_type_support(data_type::f32)
+                    && attr()->has_default_values(
+                            smask_t::oscale_runtime | smask_t::post_ops)
+                    && post_ops_ok() && attr_oscale_ok();
+            if (!ok) return status::unimplemented;
+
+            auto conf_status = acl_matmul_utils::init_conf_matmul(amp_, src_md_,
+                    weights_md_, dst_md_, bias_md_, *desc(), *attr());
+
+            if (conf_status != status::success) return status::unimplemented;
+            // Number of threads in Compute Library is set by OMP_NUM_THREADS
+            // dnnl_get_max_threads() == OMP_NUM_THREADS
+            acl_common_utils::acl_thread_bind();
+
+            return status::success;
+        }
+
+        acl_matmul_conf_t amp_;
+
+    protected:
+        bool post_ops_ok() const {
+            using namespace data_type;
+            using namespace alg_kind;
+            auto const &po = attr()->post_ops_;
+            auto is_eltwise
+                    = [&](int idx) { return po.entry_[idx].is_eltwise(); };
+            bool eltwise_only = (po.len() == 1) ? is_eltwise(0) : false;
+            bool eltwise_ok = false;
+            if (eltwise_only) {
+                const auto act_type = po.entry_[0].eltwise.alg;
+                eltwise_ok = acl_matmul_utils::acl_act_ok(act_type);
+            }
+            return eltwise_ok || (po.len() == 0);
+        }
+
+        bool attr_oscale_ok() const {
+            const auto &oscale = attr()->output_scales_;
+            return oscale.mask_ == 0;
+        }
+    };
+
+    acl_matmul_t(const pd_t *apd) : primitive_t(apd) {}
+
+    status_t create_resource(
+            engine_t *engine, resource_mapper_t &mapper) const override {
+        if (mapper.has_resource(this)) return status::success;
+        auto r = utils::make_unique<acl_resource_t>();
+        if (!r) return status::out_of_memory;
+
+        // Configure the resource based on information from primitive descriptor
+        auto st = r->configure(pd()->amp_);
+        if (st == status::success) { mapper.add(this, std::move(r)); }
+
+        return st;
+    }
+
+    typedef typename prec_traits<data_type::f32>::type data_t;
+
+    status_t execute(const exec_ctx_t &ctx) const override {
+        return execute_forward(ctx);
+    }
+
+private:
+    // To guard the const execute_forward(), the mutex must be 'mutable'
+    mutable std::mutex mtx;
```

Comment:
primitives are immutable, this is required to keep them in primitive cache, only its resources can be modified during execution. Is it possible to allocate mutex as part of resources?

Reply:
Hi @igorsafo , from what I understand this feature was added recently here https://github.com/oneapi-src/oneDNN/pull/1131 as a way to provided thread-safe access to resource_t. I would keep this inline with what added previously and then address the your comment as a separate PR for all primitives integrated with ACL.

Reply:
I see, thank you for the explanation. Resolving.
Just to make it clear, with the current approach multiple similar **independent** matmul primitives will share the same mutex (due to cache mechanism and shared primitive entry in cache) so their execution will be sequential. Once mutex is placed in resource_t, each matmul will work with its own mutex so multiple matmul primitives can be executed concurrently.

Reply:
Thanks @igorsafo, I don't think this is something we've actually considered before. As @cfRod suggested, we should take a look.